### DISCUSSION TITLE
Bunch Folders

### DISCUSSION THREAD

**Discussion by moffat** [ttscoff/bunch-docs#39] [Ideas]
Hey Brett,

As my bunches accumulate, I thought it might be nice to have folders. A 'folder' frontmatter key could be used for this. If absent (the default) the bunch would be displayed as it is today. If 'folder: work', bunch would create a 'work' folder when the user clicks on the menu bar icon. 

Best,
Carlos

**Comment by ttscoff** [comment]
That's an interesting take. I had considered using actual filesystem folders and just reading the tree in as a menu, but for my own uses there were reasons that would be a pain. 

The primary issue I see is keyboard control. The next version adds a lot of keyboard stuff, from a hotkey to display the menu to assignable shortcuts for each Bunch. Hotkeys can't open submenus, though the keyboard shortcuts assigned to the Bunch menu items should still function. You just wouldn't be able to see them. That's probably a minor consideration, but having used this keyboard-centric version for over a month now, I never click the menu anymore, but I do need to see the shortcuts displayed to remember them all, so it's not completely trivial.

This is something I've definitely considered, though, and will think it through a bit further.

**Reply by moffat** [reply]
Ah, that's a good point. For my use case (I haven't made use of hotkeys much) I'd much prefer to use the bunch itself to assign a folder than having to organize them in the filesystem, but I see you point. Looking forward to seeing you where you land on this. Thanks.

**Comment by ttscoff** [comment]
The more I think about this, the more actual filesystem folders seem easier, both on me to code, and for management of a lot of files. 

First, there'd be the risk of the frontmatter value having any misspelling or inconsistent caps, leading to unpredictable results.

Second, if you have so many Bunches that the menu needs folders, it seems at lot easier to drag items between folders than to individually edit the frontmatter for every file you want to move.

And like I said, if the files were in subfolders, it would be much easier to build the menu by just iterating through the directories rather than parsing them all then rebuilding the arrays based on frontmatter values before generating the menu. But that's doable, I already do similar with the `menu order` key.

So can you detail for me what the benefits to frontmatter over filesystem would be for you?

**Reply by moffat** [reply]
Yes, I see what you are saying, and reading over your arguments that's probably the way to go. Something I hadn't considered is that if I have a bunch of bunches with a 'folder' key and I decide to change the name of the folder, I would need to find/replace in all of them, instead of just renaming the folder. 

Are you planning to limit the depth of the folder structure? This could get unwieldy, or a user could set ~/ as the bunch directory by accident! 

Anyway, it's great that you're thinking of implementing this.

**Comment by ttscoff** [comment]
Fair point about folder depth. I would think one nested level would be 
all I would allow, anything further than that would just make unwieldy 
menus anyway.

-Brett

**Reply by apoc527** [reply]
Is "one nested level" the same as "two folders deep?" If so, I'm on board. If "one nested level" is "just 1 folder" then may I make the case for allowing Folder --> Folder --> Bunches? Now that I think about it, "one nested level" almost certainly means Folder --> Folder --> Bunch, but let me know if I'm wrong and I'll try to convince you as to why it should be allowed!

**Reply by ttscoff** [reply]
(Repositioning this comment as a reply)

 One nested level means you have your Bunch Folder, then it can contain subfolders containing more bunches. Each subfolder would create a submenu with the folder name as the title, and the menu items generated from the Bunches it contains. You just wouldn't be able to go deeper than that (Bunch Folder>Folder>Folder).

**Comment by ttscoff** [comment]
I rewrote some core components that will make a transition to this type of thing easier, but I'm going to put off implementing it for the time being. Hopefully I'll get around to it before everyone has so many Bunches they just give up :).

As a side note, I don't know if you've played with the [interactive dialog features](http://brettterpstra.com/bunch-beta/docs/bunch-files/interactivity/#multiplechoice) at all, but I've been able to combine a lot of similar Bunches into single ones that just offer a choice when opening. My "Code" Bunch, for example, combines three different Bunches that used to launch specific things for "Coding Bunch," "Coding Marked," and "Coding nvUltra." Now I just launch "Code" and it asks me what I'm working on, Bunch, Marked, or nvUltra, then launches the appropriate apps and folders for each one. This is layed out in [an example Bunch](http://brettterpstra.com/bunch-beta/docs/bunch-files/samplebunch/#coding) if you're curious.

**Reply by moffat** [reply]
I like the idea of the prompt a lot. Thanks for the suggestion.

**Reply by ttscoff** [reply]
It really is handy. I'll probably have to keep reminding people about
the flexibility of the interaction prompts when combined with snippets,
but there are a ton of possibilities. Truly interactive snippets, and
for the most part it's all very keyboard-centric. The dialogs should
always take focus and can be navigated with arrows, space, return, and
escape. Let me know if you start playing around with it and have any
questions/ideas.

**Reply by apoc527** [reply]
Any chance you could prepare an example Bunch that's a bit simpler, maybe set up with apps and folders that everyone would have? I'm struggling to make use of the interactive dialogue feature...but I also haven't tried TOO hard quite yet simply because of the...shall we say "rapid"...recent pace of development (I'm waiting, in other words). :-)

**Reply by apoc527** [reply]
LOL I tried to run the sample bunch and that did not work. To be fair, I knew it wouldn't work, but FYI, it crashed Bunch Beta. Sadly it crashed before I could see any dialog options too.

**Reply by ttscoff** [reply]
Which sample bunch. The one that it automatically creates when it installs (which should only run apps everyone has) or one from the Sample Bunches page?

**Reply by apoc527** [reply]
The one you linked to above. It's your "Code" bunch, I think.

**Reply by ttscoff** [reply]
If you can find the crash report from that crash, would be helpful to take a look at it. If everything goes the way it should, Bunch should never crash, even with a Bunch that's completely wrong for your system.

**Reply by ttscoff** [reply]
> The one you linked to above. It's your "Code" bunch, I think.

That one's kind of intense. I'd love to know which part of it caused a crash for you. Could you set your Log Level to Debug (Help->Log Level) and open the log when you try running it? I'd love to see the output, in addition to any crash report.

**Reply by apoc527** [reply]
Well it didn't crash this time and you just about made me jump clear through the ceiling when the robotic voice loudly announced that the Bunch launched, but here are the logs if still relevant!
[com.brettterpstra.BunchBeta 2021-03-26--16-24-19-778.log](https://github.com/ttscoff/bunch/files/6213107/com.brettterpstra.BunchBeta.2021-03-26--16-24-19-778.log)

In other news, I'm glad I found these logs, super helpful to...umm...debug my bunches. :-)

**Reply by ttscoff** [reply]
Thanks. If everything worked, great... doesn't help me track down the original crash, but I do always like it when stuff works.

**Reply by moffat** [reply]
> It really is handy. I'll probably have to keep reminding people about
> the flexibility of the interaction prompts when combined with snippets,
> but there are a ton of possibilities. Truly interactive snippets, and
> for the most part it's all very keyboard-centric. The dialogs should
> always take focus and can be navigated with arrows, space, return, and
> escape. Let me know if you start playing around with it and have any
> questions/ideas.

This worked really well for me. By the way, thanks for the detailed documentation. I couldn't make this work last night and of course I needed to RTFM.

**Comment by ttscoff** [comment]
Ok, I've decided to experimentally add this (I got up too early this morning). As it stands, any subfolders in the main Bunch folder will be searched for `.bunch` files. If any are found, a submenu with the same name as the subfolder is added at the bottom of the Bunch list, containing a menu item for each Bunch in the subfolder.

Caveats:

- Automatic shortcut keys (1-9, etc.) are disabled for submenus. Any custom shortcuts added via frontmatter are respected.
- All relative paths used in Bunches are relative to the base Bunch Folder, not to the subfolder. So if you run `$ myscript.sh` from a Bunch in a subfolder, it's going to look for `myscript.sh` in the main Bunch folder. You would have to include `subfolder/myscript.sh`
- `.startup` scripts in subfolders aren't used

Do those caveats sound reasonable? If so, this will be working in the next update.

I'm also wondering if this should be a preference or not, but it seems like intuitive enough behavior that it could just be enabled for everyone...

**Reply by moffat** [reply]
I don't make use of the shortcut keys myself, so this doesn't impact me. The other caveats sound very reasonable to me. I agree you don't need a preference for this, the documentation should be enough.

**Reply by ttscoff** [reply]
I can't imagine using Bunch without the keyboard shortcuts now that it has them. So you still click the menu every time?

**Reply by ttscoff** [reply]
BTW, I've added support for a `folder.frontmatter` file inside of subfolders that lets you either completely ignore the folder, or set the display title and menu divider (before|after) for the submenu. I left it open-ended so I can add additional frontmatter keys as they become necessary.

This will all be in the next update after a little more local testing.

**Reply by moffat** [reply]
For now I am just clicking the menu, but I could definitely see switching to the shortcuts soon.

**Reply by moffat** [reply]
> BTW, I've added support for a `folder.frontmatter` file inside of subfolders that lets you either completely ignore the folder, or set the display title and menu divider (before|after) for the submenu. I left it open-ended so I can add additional frontmatter keys as they become necessary.
> 
> 
> 
> This will all be in the next update after a little more local testing.

Great!

**Comment by ttscoff** [comment]
Ok, this is implemented in b82, which is out now. Please see [the documentation](https://brettterpstra.com/bunch-beta/docs/using-bunch/organizing-bunches/#sort-with-submenus) for details.

**Reply by moffat** [reply]
Awesome, thanks!

**Comment by stuzog** [comment]
Great improvement! My bunches list was getting long. Love the new bunch Folders but don't like how bunches in a folder by default listed as folder/bunch when folder clicked. I know where the bunches are and this makes the bunch title harder to read. I would prefer just the bunch titles to show in folder list without having to add title frontmatter to each bunch file. Not just me being lazy but this would make it easier to move bunch files between folders without editing each one.

**Reply by ttscoff** [reply]
Ok, this is fixed for the next update. I didn't notice it because it doesn't happen if you have `title:` frontmatter in the Bunch, which is always the first thing I add :).

**Comment by ttscoff** [comment]
Oh, my bad. That definitely wasn't intentional. Will fix for the next 
update.

-Brett
